src/TestProblem/Hydro/CCSN/Record_CCSN.cpp

```diff
@@ -0,0 +1,309 @@
+
+//    file header
+      if ( FirstTime )
+      {
+         if ( Aux_CheckFileExist(filename_central_dens) )
+         {
+             Aux_Message( stderr, "WARNING : file \"%s\" already exists !!\n", filename_central_dens );
+         }
+
+         else
+         {
+             FILE *file_max_dens = fopen( filename_central_dens, "w" );
+             fprintf( file_max_dens, "#%14s %12s %16s %16s %16s %16s\n",
+                                     "Time [sec]", "Step", "Dens [g/cm^3]", "PosX [cm]", "PosY [cm]", "PosZ [cm]" );
+             fclose( file_max_dens );
+         }
+
+         FirstTime = false;
+      }
+
+      FILE *file_max_dens = fopen( filename_central_dens, "a" );
+      fprintf( file_max_dens, "%15.7e %12ld %16.7e %16.7e %16.7e %16.7e\n",
+               Time[0]*UNIT_T, Step, DataCoord[0]*UNIT_D, DataCoord[1]*UNIT_L, DataCoord[2]*UNIT_L, DataCoord[3]*UNIT_L );
+      fclose( file_max_dens );
+
+   } // if ( MPI_Rank == 0 )
+
+} // FUNCTION : Record_CCSN_CentralDens()
+
+
+
+#ifdef GRAVITY
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Record_CCSN_GWSignal
+// Description :  Record the second-order time derivative of mass quadrupole moments
+//
+// Note        :  1. Invoked by Record_CCSN()
+//                2. Ref: Kenichi Oohara, et al., 1997, PThPS, 128, 183 (arXiv: 1206.4724), sec. 2.1
+//-------------------------------------------------------------------------------------------------------
+void Record_CCSN_GWSignal()
+{
+
+   const char   filename_QuadMom_2nd[ ] = "Record__QuadMom_2nd";
+   const double BoxCenter           [3] = { 0.5*amr->BoxSize[0], 0.5*amr->BoxSize[1], 0.5*amr->BoxSize[2] };
+
+// allocate memory for per-thread arrays
+#  ifdef OPENMP
+   const int NT = OMP_NTHREAD;   // number of OpenMP threads
+#  else
+   const int NT = 1;
+#  endif
+
+   const int NData   = 6;
+   const int ArrayID = 0;
+   const int NPG_Max = POT_GPU_NPGROUP;
+
+   double   QuadMom_2nd[NData] = { 0.0 };
+   double **OMP_QuadMom_2nd    = NULL;
+   Aux_AllocateArray2D( OMP_QuadMom_2nd, NT, NData );
+
+
+#  pragma omp parallel
+   {
+#     ifdef OPENMP
+      const int TID = omp_get_thread_num();
+#     else
+      const int TID = 0;
+#     endif
+
+//    initialize arrays
+      for (int b=0; b<NData; b++)   OMP_QuadMom_2nd[TID][b] = 0.0;
+
+      for (int lv=0; lv<NLEVEL; lv++)
+      {
+         const double dh = amr->dh[lv];
+         const double dv = CUBE( dh );
+
+         const double TimeNew   = Time[lv];
+         const int    NTotal    = amr->NPatchComma[lv][1] / 8;
+               int   *PID0_List = new int [NTotal];
+
+         for (int t=0; t<NTotal; t++)  PID0_List[t] = 8*t;
+
+
+         for (int Disp=0; Disp<NTotal; Disp+=NPG_Max)
+         {
+            int NPG = ( NPG_Max < NTotal-Disp ) ? NPG_Max : NTotal-Disp;
+
+            Prepare_PatchData( lv, TimeNew, &h_Pot_Array_P_Out[ArrayID][0][0][0][0], NULL,
```

Comment:
I didn't realize nested OpenMP is disabled in GAMER. I rearrange the structure so that `omp parallel` is applied after `Prepare_PatchData()` as you suggested.